### DISCUSSION TITLE
Macro definitions expanding across multiple lines

### DISCUSSION THREAD

**Discussion by hroncok** [rpm-software-management/rpm#2353] [Ideas]
Hello. I often written macros defined like this:

```rpm
%py_shebang_fix %{expand:\\\
  if [ -z "%{?py_shebang_flags}" ]; then
    shebang_flags="-k"
  else
    shebang_flags="-ka%{py_shebang_flags}"
  fi
  %{__python} -B %{_rpmconfigdir}/redhat/pathfix.py -pni %{__python} $shebang_flags}
```

I've abused `%expand` to have a macro definition that expands across multiple lines.

@pmatilai [said](https://bugzilla.redhat.com/show_bug.cgi?id=2160716#c10):

> Using an outer `%{expand:...}` just to make it look like a function (as seems to be a common habit in Fedora) is bad thing to do because it then breaks when you least expect it to. NEVER use `%{expand:...}` unless you actually mean it, as in, need that extra expand. And even then, only use it on the smallest possible block to achieve what you need, not the entire thing "just because"

If I had to drop the  `%expand`, I'd probably have to do something like:

```rpm
%py_shebang_fix \
  if [ -z "%{?py_shebang_flags}" ]; then \
    shebang_flags="-k" \
  else \
    shebang_flags="-ka%{py_shebang_flags}" \
  fi \
  %{__python} -B %{_rpmconfigdir}/redhat/pathfix.py -pni %{__python} $shebang_flags
```

That is a bit **ugly**. Another option is to take a macro that is always defined and make the thing conditional on it:

```rpm
%py_shebang_fix %{?nil:\\\
  if [ -z "%{?py_shebang_flags}" ]; then
    shebang_flags="-k"
  else
    shebang_flags="-ka%{py_shebang_flags}"
  fi
  %{__python} -B %{_rpmconfigdir}/redhat/pathfix.py -pni %{__python} $shebang_flags}
```

That is **hard to understand** by the reader.

Could we get some kind of macro that allows us to achieve what we use `%expand` for?

E.g.

```rpm
%py_shebang_fix %{multiline:\
  if [ -z "%{?py_shebang_flags}" ]; then
    shebang_flags="-k"
  else
    shebang_flags="-ka%{py_shebang_flags}"
  fi
  %{__python} -B %{_rpmconfigdir}/redhat/pathfix.py -pni %{__python} $shebang_flags}
```

**Comment by pmatilai** [comment]
Yup, the best way to get people off a bad habit is to offer something better :smile:

**Comment by pmatilai** [comment]
Probably the sanest/most intuitive alternative would be just let people use { } to indicate such blocks.
That actually already works:
> [pmatilai🎩︎localhost rpm]$ rpm --define "foo { aa bb }" --eval %foo
 aa bb 

... but within one line only :facepalm: Over multiple lines it fails with "macro has an unterminated body"

**Comment by mlschroe** [comment]
I always considered this not working to be a bug.

**Comment by mlschroe** [comment]
It's not hard to fix this for the macro file case. But the spec file case is a bit nastier.

**Reply by pmatilai** [reply]
The more complicated macros tend to be in macro files, whereas specs more commonly have just one-liners to flip stuff on/off, redefine a value here or there etc. Fixing this for the macro file case would improve life quality for many, regardless of whether it works in specs or not. Patches most welcome :grin:
